오픈마켓 II [STEP 2] 웡빙, 보리사랑

yusw10 opened the pull request:
# 오픈마켓2 STEP2 PR

###### tags: `PR`
@ehgud0670 
@wongbingg
@yusw10

안녕하세요 제이슨! 
오픈마켓 2의 STEP2 PR 보냅니다...!
마지막 까지 잘 부탁드립니다🤗

## UML
### Class Diagram(Step3 PR 기준)

<img src=https://i.imgur.com/84QSQmP.png>

## Step 4 기능구현

### 1. 상품 상세 페이지 UI구현
<img src=https://i.imgur.com/ikmwPUW.jpg width=70%>

- ImageView Paging
- Paging Index

### 2. 액션시트 구현
디테일 뷰에서 오른쪽 위의 액션버튼 탭시 수정/삭제 두 가지의 액션을 선택하도록 구현 
- 수정버튼 탭시 : 이전에 구현해놓았던 "상품수정" 창으로 이동
    - 텍스트 필드에 채워져 있는 기존의 정보를 변경할 수 있다
    - 변경 후 우측상단의 Done 버튼 탭시 HTTP PATCH 메서드를 실행시킨 후, 상품 수정 완료 Alert
    - 수정 완료 Alert 에서 확인버튼 탭시 이전 화면인 ProductDetailView로 이동
    - (ProductDetailView에서 수정된 사항이 잘 적용되었는지 확인가능)
- 삭제버튼 탭시 : alert로 삭제시 필요한 secretKey를 textField 에 요구
    - secretKey가 일치할 시, HTTP DELETE 메서드를 실행시킨 후, 삭제 성공 Alert
    - 삭제성공 Alert 에서 확인버튼 탭시 MainView로 이동
    
### 3. Infinite Scrolling 구현 
- 스크롤이 바닥에 닿았을 때, 다음 데이터를 20개씩 받아오도록 구현

### 4. 새로고침 구현
- refreshControl을 만들어 위로 당겼을 때, 최신 데이터의 첫 페이지를 받아오도록 구현

### 5. 상품 상세페이지의 이미지 페이징 구현
- 등록된 사진의 이미지 갯수와 각 이미지의 크기에 맞게 페이징 구현
- 몇번째 사진을 보고있는지 페이징 인덱스를 이미지 아래에 출력되도록 구현

## 궁금한 점 / 고민한 점

### 같은 데이터를 공유하는 두개의 화면 구현방법
저희는 하나의 Collection View에서 List와 Grid 형태의 레이아웃, List와 Grid만을 위한 DiffableDataSource를 각각 만들었습니다(한개의 콜렉션뷰, 두개의 레이아웃, 두개의 데이터 소스) 같은 데이터를 공유하기에 이를 표현하는 콜렉션뷰는 한개가 맞지만 형태가 다르기에 데이터를 표현할 데이터소스는 각자 존재하는게 맞다고 생각했기에 이렇게 구현해보았습니다.

또 다른 캠퍼들이 구현한 사례들을 보면. 한개의 콜렉션뷰 한개의 데이터소스에 레이아웃만 교체하는 방식(셀 내부에 셀의 레이아웃을 변경하는 코드가 있었습니다) 도 있었고, 아예 독립된 콜렉션뷰로 따로 만들어 isHidden 속성으로 번갈아 가며 화면에 출력하는 방식도 있었습니다.

각각 일장 일단이 있다고 생각하는데 제이슨은 **같은 데이터를 출력하되 다른 레이아웃으로 표현해야 한다면 어떤식으로 구현하시는지 궁금했습니다.**

### 사전에 협의되지 않는 응답(?)에 대하여
밑의 사진처럼 상품을 등록할 때(POST) 필요한 params를 보면 descriptions가 필수로 포함되어야 한다고 되어있는걸 확인 했습니다.
<img src=https://i.imgur.com/xV3fxuy.png width=90%>

그래서 저희는 서버에서 상품상세 정보를 요청하는 HTTP 메서드의 반환값으로 반드시 description을 받도록 구현했습니다. 

<img src=https://i.imgur.com/iPcoF28.png width=80%>

실제로 위와같이 응답데이터에도 description이 포함되있는것을 확인했고 아래와 같이 정상응답을 기준으로 코드를 구현했습니다.

<img src=https://i.imgur.com/5NcqvnY.png width=30%>

문제는 수정을 구현하면서 상품 상세정보를 받아오는 과정에서 일어났는데, description이 없는 데이터가 넘어오는 경우가 있었습니다(다른 캠퍼가 postman 과 같은 프로그램으로 데이터를 넣지 않고 request를 보냈는데 정상 등록된 것 같습니다) 

> 중간에 price와 bargain_price사이에 description이 빠져있는걸 볼 수 있습니다.

<img src=https://i.imgur.com/26rDNpq.png width=50%>

이렇게 서버 API 문서를 기반으로 구현을 했는데 안넘어오는 경우가 생겨서 현재는 description이 nil로 넘어올 경우 공백 문자열을 강제로 넣어주는 조건문을 넣어준 상황입니다.

이런 경우는 서버를 담당하시는 분의 문제인걸까요? 보통 어떻게 풀어가시는지 궁금했습니다. 또 현재 서버에 저런 데이터가 이미 몇몇 존재하는데 지금처럼 조건문으로 해결해두어도 괜찮은지도 궁금합니다.

## 트러블 슈팅!

### MVC로 다시 리팩토링
STEP1이 끝나고 전체적인 코드를 확인해보니, 제이슨이 말하신대로 MVC 아키텍쳐에 맞게 구현되어 있지 않는것을 확인했습니다. 특히 메인화면의 상품 리스트 데이터를 뷰 컨트롤러가 직접 소유하며 호출하고 수정하기에 모델과 컨트롤러의 구분이 모호했었습니다.

전체적인 구조를 리팩토링하기 위해 서로 MVC를 구현하는 흐름에대해 다시 공부하였고, 상품의 리스트를 컨트롤러로부터 분리하여 담당하는 `ProductListManager`타입을 구현하여 실제 상품의 배열과 컨트롤러를 분리하도록 리팩토링을 진행하였습니다.

wongbingg committed:
feat: 새로고침을 하여 최신 페이지를 불러오는 refreshControl 구현

- 메인화면에서 최상단으로 가서 화면을 아래방향으로 당길 때, Activity Indicator가 돌아가며 새로운 20개의 상품을 불러오도록 구현
- 수정 화면에서 포스트를 완료하고 메인화면으로 돌아올 때, Notification을 전달해주어 refresh가 되도록 구현

wongbingg committed:
refactor: 메서드 네이밍 변경 및 snapshot 등록을 조건문으로 분기

- productManager 내의 fetch 데이터의 네이밍을 update 로 변경
- 2개의 데이터소스가 세그먼트의 상태에 따라 다르게 적용 되도록 snapshot 등록
- list 에서 grid로 넘어갈 때, 스크롤 했던 위치를 기억하기 위해 scrollToItem 메서드 사용

wongbingg committed:
feat: 아래로 내리면서 20개의 상품씩 계속 추가되도록 paging 구현

- collectionView 스크롤이 아래 끝까지 내려갔을 때 현재 페이지 수에 1을 추가하여 다시 데이터를 로드하도록 구현
- 이전에 있던 상품에 새로운 20개의 상품이 추가되어 모델에 저장
- 모델이 바뀌면서 MainViewController으로 notification을 보내 datasource등록을 시킴

yusw10 committed:
feat: 상품 상세 뷰 컨트롤러 구현

- 수정 및 삭제 액션시트 구성
- 상품의 상세 정보 저장을 위한 뷰 컨트롤러 구현

yusw10 committed:
refactor: Snapshot 구성 로직 변경

- list와 grid가 snapshot을 서로 주입하며 동기화 할 수 있도록 로직 변경

yusw10 committed:
feat: ProductSetupViewController 가 수정과 등록에 따라 다르게 작동하도록 수정

- 상품 상세에서 넘어올 경우 이미지 변경이 불가하고 세부 내용을 변경할 수 있도록 구현
- 등록화면일 경우 PickerImageViewController를 활용하여 이미지를 등록할수 있게 구현
- description에서 빈 문자열을 처리할 수 있도록 예외처리

wongbingg committed:
feat: ProductDetailViewController 에서 HTTP-DELETE 구현

- 삭제시 필요한 secret키가 유효하다면, HTTP-DELETE 메서드를 실행하도록 구현

wongbingg committed:
chore: 컨벤션 수정

- @objc -> private -> Internal 메서드 순으로 정렬
- 불필요한 개행 제거

wongbingg committed:
refactor: segment 로직 수정

- datasource 끼리 들고있는 데이터가 달라서 생긴 충돌 오류를 해결하기 위해 조건문으로 분리

yusw10 committed:
feat: 상품 상세 화면 구현

- 상품 클릭시 상품 상세 정보를 표현하는 화면 구성
- 상세화면의 이미지가 페이징이 되도록 구현

wongbingg committed:
feat: DetailView의 세부사항 추가

- scroll 메서드의 시점을 이용해 사진에 따라 페이지 수를 출력하여 pagingLabel에 적용
- 할인가격을 적용해주는 setPriceLabel 메서드 적용
- productName 의 폰트를 bold체로 변경

yusw10 committed:
docs: README.md Update

wongbingg committed:
docs: ReaMe 수정

ehgud0670 commented:
# 궁금한 점 / 고민한 점

## 1. 같은 데이터를 출력하되 다른 레이아웃으로 표현해야 한다면 어떤 식으로 구현하시는지 궁금했습니다.

> 저희는 하나의 Collection View에서 List와 Grid 형태의 레이아웃, List와 Grid만을 위한 DiffableDataSource를 각각 만들었습니다(한개의 콜렉션뷰, 두개의 레이아웃, 두개의 데이터 소스) 같은 데이터를 공유하기에 이를 표현하는 콜렉션뷰는 한개가 맞지만 형태가 다르기에 데이터를 표현할 데이터소스는 각자 존재하는게 맞다고 생각했기에 이렇게 구현해보았습니다.

=> 저는 해당 코드를 보고 되게 좋다고 생각을 했습니다. collectionView는 하나만을 사용하지만 서로 다른 레이아웃을 표현하기 위해 두 개의 레이아웃, 두 개의 데이터소스가 각각 존재하기 때문에 유지보수하기 좋은 구현이라고 생각했기 때문입니다.

> 아예 독립된 콜렉션뷰로 따로 만들어 isHidden 속성으로 번갈아 가며 화면에 출력하는 방식도 있었습니다.

=> 저는 두 개의 콜렉션뷰를 쓰고 isHidden을 사용하는건 개인적으로 어색하다고 생각이 들었습니다. 구현을 한다면 isHidden을 사용하지 않고PageViewController 아래 두 VC를 만들고 각각의 VC에 한개씩 콜렉션 뷰를 사용할 것 같습니다. 

뷰 구현에 정답은 없기 때문에 참고만 하시면 될 듯 합니다 :)

## 2. 이런 경우는 서버를 담당하시는 분의 문제인걸까요? 보통 어떻게 풀어가시는지 궁금했습니다. 또 현재 서버에 저런 데이터가 이미 몇몇 존재하는데 지금처럼 조건문으로 해결해두어도 괜찮은지도 궁금합니다.

서버에서 필수 값을 보내야 하는데 보내지 않은 건 엄연히 서버 잘못이지만 충분히 생길 수 있는 문제 상황입니다.
예를 들어 이미 배포된 서버 응답값에는 클라이언트가 잘 대응하다가 새로 서버를 배포했을 때 서버 응답값이 잘못 내려와 클라이언트 쪽에서 디코딩을 못해 아무런 화면을 보이지 못한다면 사용자는 아무것도 못볼 것이기 때문입니다. 
따라서 어느 경우라도 디코딩을 잘해 일부 값이라도 보이기 위해 **DTO 객체의 프로퍼티로 옵셔널 타입을 쓰고 타입은 Enum 대신 raw한 타입을 쓰는(ex. `String`, `Int`...)방법**을 쓸 수 있습니다. 필수 값임에도 불구하고 내려오지 않거나 잘못된 값이 내려오는 상황을 대비해 옵셔널을 사용하고 Enum 타입 사용을 지양하는 것이죠!

> 예시

```swift
struct OnethingUserModel: Codable {
    let name: String?
    let email: String?
    let authType: String?
    let nickname: String?
    let imageType: String?
    let enableAlarm: String?
    
    var castingAccessType: SocialAccessType? {
        guard let authType = self.authType,
              let accessType = SocialAccessType(rawValue: authType)
        else { return nil }
        
        return accessType
    }
    
    var profileImageType: OnethingProfileType? {
        guard let strongImageType = self.imageType,
              let imageType = OnethingProfileType(rawValue: strongImageType)
        else { return nil }
        
        return imageType
    }
    
    var castingAlarm: AlarmToggle? {
        guard let enableAlarm = self.enableAlarm,
              let alarmToggle = AlarmToggle(rawValue: enableAlarm)
        else { return nil }
        
        return alarmToggle
    }
    
    enum AlarmToggle: String {
        case on = "ON"
        case off = "OFF"
    }
}
```
=> 이 예시는 필수값이어도 Optional 로 받고 Enum화 할수 있는 프로퍼티여도 raw한 타입으로 선언하고 이후에 이 모델값을 사용할때는 연산프로퍼티를 통해 Enum 생성후 이용하는 방식입니다.


질문은 답변 드렸고 코드 리뷰 바로 드리도록 하겠습니다 :)

ehgud0670 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -76,13 +99,44 @@ final class MainViewController: UIViewController {
-    // MARK: - Main View Controller Method
+    // MARK: - @objc method
+    @objc private func addButtonDidTapped() {
+        print("add button tapped")
```
* 로그가 필요한 정보들이라면 printLog 와 같이 따로 함수를 구현해 이건 콘솔 로그로 보여야 할 정보여서 print를 한다는 식으로 일반 print문과 구분해주시면 코드가 더 깔끔해보이고 좋을 것 같아요.
* 만약 디버깅을 위해 print를 한 것이라면 print문을 제거해주시고 breakpoint 로 디버깅하시면 좋을 듯 합니다 :)

yusw10 replied:
말씀 주신대로 디버깅을위해 한것이라 print 기능을하는 부분을 fatalError를 사용하도록 수정해보았습니다. 

[4102255](https://github.com/yagom-academy/ios-open-market/pull/201/commits/4102255fce6882860e4c24ced40f675b34e759f9)

ehgud0670 replied:
fatalError는 앱 크래시가 반드시 발생하므로 해당 구간에 사용하면 위험할 듯 합니다.
fatalError를 쓰신 부분은 꼭 디버깅이 아니고 로깅을 위해서도 print가 필요한 곳이네요!
위에 말씀 드린 것 처럼 해당 부분처럼 꼭 print 가 필요한 부분이라면 pringLog 와 같은 함수만 구현해서 로그임을 구분만 시켜주면 될 듯 합니다 :)  

> printLog 단순 구현 예 

```swift
func printLog(message: String) {
    print(message)
}
```

ehgud0670 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -16,27 +16,47 @@ final class MainViewController: UIViewController {
     private var gridLayout: UICollectionViewLayout? = nil
     private var productListManager = ProductListManager()
     private var currentMaximumPage = 1
+    private var refresher: UIRefreshControl!
```
이건 이제 개인의 취향이라고 여기고 넘길게요 ㅎㅎ 🥲

yusw10 replied:
앗... 강제 언래핑이나 암시적 추출을 최대한 지양하면서 코드 작성하려고 했는데 UIRefreshControl에 대한 지식이 부족하다보니 예제 코드를 활용하면서 나온 실수 인것같습니다...! 이부분도 수정해보았습니다.

[bde13e7](https://github.com/yagom-academy/ios-open-market/pull/201/commits/bde13e7be7d4f768a7e083216c0b7c5659bf6efd)

ehgud0670 replied:
좋습니다 👍

ehgud0670 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -16,27 +16,47 @@ final class MainViewController: UIViewController {
         case main
     }
     
     private var shouldHideListLayout: Bool? {
```
shouldHideListLayout는 언제든 쓰이는 값이고 true 아니면 false 일 것 같은데 논옵셔널타입이 더 적절해 보입니다 :)

yusw10 replied:
넵 말씀주신대로 `shouldHideListLayout `에 기본값을 부여 하는 방향으로 수정해보았습니다.

[1827bd6](https://github.com/yagom-academy/ios-open-market/pull/201/commits/1827bd6cb6598e31e836c902ccb99adcaad69628)

ehgud0670 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -16,27 +16,47 @@ final class MainViewController: UIViewController {
+                    collectionView.setCollectionViewLayout(gridLayout, animated: true) { _ in
+                        self.loadData()
+                    }
+                } else {
+                    gridDataSource?.apply(listSnapShot)
+                    collectionView.dataSource = gridDataSource
+                    collectionView.setCollectionViewLayout(gridLayout, animated: true) { _ in
+                        DispatchQueue.main.async {
+                            self.collectionView.reloadData()
+                        }
+                    }
                 }
-                collectionView.dataSource = gridDataSource
-                collectionView.setCollectionViewLayout(gridLayout, animated: true)
             } else {
-                
-                guard let listLayout = listLayout else {
-                    return
+                if gridSnapShot.numberOfItems < listSnapShot.numberOfItems {
+                    collectionView.dataSource = listDataSource
+                    collectionView.setCollectionViewLayout(listLayout, animated: true) { _ in
+                        self.loadData()
+                    }
+                } else {
+                    listDataSource?.apply(gridSnapShot)
+                    collectionView.dataSource = listDataSource
+                    collectionView.setCollectionViewLayout(listLayout, animated: true) { _ in
+                        DispatchQueue.main.async {
+                            self.collectionView.reloadData()
+                        }
+                    }
                 }
```
* 중복이 보이는데 메소드로 빼서 처리해주시면 좋을 듯 합니다! **DRY(Don't Repeat Yourself)**

wongbingg replied:
중복으로 쓰이는 로직들을 메서드 분리 해주었습니다 !

ehgud0670 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -16,27 +16,47 @@ final class MainViewController: UIViewController {
             guard let shouldHideListLayout = shouldHideListLayout else { return }
-            print("DID TAPPED SEGMENT CONTROLLER")
+            guard let gridLayout = gridLayout, let listLayout = listLayout else { return }
+            guard let gridSnapShot = gridDataSource?.snapshot(),
+                  let listSnapShot = listDataSource?.snapshot() else { return }
             if shouldHideListLayout {
-                guard let gridLayout = gridLayout else {
-                    return
+                if gridSnapShot.numberOfItems > listSnapShot.numberOfItems {
+                    collectionView.dataSource = gridDataSource
+                    collectionView.setCollectionViewLayout(gridLayout, animated: true) { _ in
+                        self.loadData()
```
didSet 내용은 메소드 한줄로 빼는게 깔끔합니다.

wongbingg replied:
말씀주신대로 , 메서드 분리를 하여 한줄로 표현되도록 수정했습니다 🧐🧐

ehgud0670 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -95,48 +149,29 @@ final class MainViewController: UIViewController {
     }
     
-    private func loadData() {
+    private func addData() {
```
addData 라기 보다는 데이터를 더 fetch 하는 것이므로 fetchMoreData 라는 식으로 네이밍하는게 좋아보입니다 :)

wongbingg replied:
더 좋은 네이밍인것 같아 수정하였습니다! ㅎㅎ

ehgud0670 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -207,15 +252,45 @@ extension MainViewController {
+            print("앙 바닥에 닿았당")
+            DispatchQueue.main.async { [weak self] in
+                print("다음 페이지")
+                self?.currentMaximumPage += 1
```
currentMaximumPage 도 모델값이므로 ProductListManager가 들고 있는게 좋을 것 같아요~

wongbingg replied:
말씀해주신대로 모델값이라 판단되어 ProductListManager에서 관리하도록 이동시켜 주었습니다. 접근제어자를 internal 로 하여 컨트롤러에서 직접 건드릴 수 있도록 만들어 주었는데, private 으로 처리하고 메서드를 통해 간접적으로 접근시켜 주는 것이 더 좋은 방법 일까요 ?

ehgud0670 replied:
private 으로, 아니면 private(set)으로 처리하고 메서드를 통해 간접적으로 해당 변수에 접근하는 것이 
함부로 해당 변수를 접근할 수 없어 유지보수 측면에서 더 좋다고 생각합니다 👍

ehgud0670 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -207,15 +252,45 @@ extension MainViewController {
+                print("다음 페이지")
+                self?.currentMaximumPage += 1
+                self?.addData()
+                self?.collectionView.reloadData()
```
저번에 말씀드렸던 것 처럼 reloadData는 모델을 옵저빙하다가 변화를 감지하면 그때 호출해주시면 될 듯 합니다! 🙂

ehgud0670 replied:
이처럼 추가 데이터를 받아올때는 뷰 쪽에서는 snapShot으로 reloadItems 말고도 insertItems 메소드를 활용하는 것도 좋을 듯 합니다 :)

yusw10 replied:
넵 현재 `ProductListManager`에서는 리스트값의 변경이 있을 경우(스크롤 해서 상품이 추가되는 경우)를 post하고 있습니다. 이를 옵저빙하여 실행되는 메서드부분에 말씀하신대로 reload 하도록 변경 해보았습니다.

[8983657](https://github.com/yagom-academy/ios-open-market/pull/201/commits/89836578c1d1fabffd5562d6ec60aaf61975c7cb)

ehgud0670 commented on OpenMarket/OpenMarket/Model/ProductListManager.swift:
```diff
@@ -8,20 +8,24 @@
 
 class ProductListManager {
-    var productList: [Product] {
+    private var productList: [Product] {
```
private 으로 선언하고 getCurrentList()으로 반환해서 사용하는 것은 그다지 의미가 더 있진 않아요! 
private 으로 선언하고 메소드로 표현해주시면 될 듯 합니다 :)
ex) 특정 인덱스의 모델을 사용하고 싶은 경우 `product(at index)`를 구현해봅니다.

ehgud0670 replied:
정정합니다. 외부에서 snapShot을 만들려면 그대로 필요하겠네요.
저는 snapShot까지 ProductListManager에서 반환해도 좋다고 생각하지만,
UIKit를 임포트하게 되는 것이 꺼려지시면 그냥 getCurrentList() 메소드를 사용하시면 될 것 같아요~

yusw10 replied:
아하... 스냅샷까지 `ProductListManager`가 관리해주는 것 까지는 생각을 못해본것 같습니다. 참고해보겠습니다...!

ehgud0670 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -207,15 +252,45 @@ extension MainViewController {
-        prodcutDetailVC.productId = productListManager.productList[indexPath.row].id
-        prodcutDetailVC.viewControllerTitle = "상품 수정"
-        print("\(productListManager.productList[indexPath.row].id) - \(productListManager.productList[indexPath.row].name) is tapped")
+        let prodcutDetailVC = ProductDetailViewController()
+        let seletedProduct = productListManager.getCurrentList()[indexPath.row]
+        prodcutDetailVC.productId = seletedProduct.id
+        prodcutDetailVC.viewControllerTitle = seletedProduct.name
+        print("\(seletedProduct.id) - \(seletedProduct.name) is tapped")
         navigationController?.pushViewController(prodcutDetailVC, animated: true)
```
detailVC로 가는 것도 메소드 분리해주시면 좋을 것 같아요 :)

yusw10 replied:
말씀주신대로 분리 해보았습니다. 
확실히 `didSelectItemAt`에서 바로 `pushDetailViewController`메서드를 호출하니 코드가 간결해진것 같아서 가독성이 좋아진다는 느낌을 받을 수 있었습니다.

[582e666](https://github.com/yagom-academy/ios-open-market/pull/201/commits/582e6661c7f412c5dca8b96442a456bcd3ab5ad1)

ehgud0670 commented on OpenMarket/OpenMarket/Controller/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,134 @@
+            imageView.contentMode = .scaleAspectFit
+            productDetailView.horizontalStackView.addArrangedSubview(imageView)
+        }
+        productDetailView.productNameLabel.text = detail.name
+        productDetailView.stockLabel.text = "남은수량 : \(detail.stock)"
+        productDetailView.descriptionTextView.text = detail.description
+        productDetailView.setupPriceLabel(currency: detail.currency, price: detail.price, bargainPrice: detail.bargainPrice)
```
이런것도 이제 하나의 메소드로 만들어 처리하면 깔끔하고 좋습니다. 

> example

* `productDetailView.update(with: model)`
* `productDetailView.update(with: viewModel)`

wongbingg replied:
View 자체적으로 update를 시키는 메서드를 만들어 놓으니 사용하기 편리한 것 같습니다! ㅎㅎ ☺️

ehgud0670 commented on OpenMarket/OpenMarket/Controller/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,134 @@
+            productDetailView.horizontalStackView.addArrangedSubview(imageView)
+        }
+        productDetailView.productNameLabel.text = detail.name
+        productDetailView.stockLabel.text = "남은수량 : \(detail.stock)"
```
```suggestion
        productDetailView.stockLabel.text = detail.stockLabelText
```
이렇게 따로 메소드나 연산프로퍼티로 만들어 뷰에 값 할당하면 재사용성이 늘어나 좋습니다. 
이러면 Presenter의 역할을 하므로 detail은 뷰모델의 역할을 하게된 셈입니다(따지자면 데이터바인딩까지 해야 뷰모델입니다).

yusw10 replied:
오 이렇게 사용하는 방법도 생각해보지 못했는데 정말 좋은것 같습니다. 
그리고 궁금한점이, 이렇게 사용되는 text값들을 각각의 모델이 연산프로퍼티로 소유하는것이 아니라 Namespace로 묶어서 사용하는 방법은 별로 좋지 않은 방법일까요?

[8ce1738](https://github.com/yagom-academy/ios-open-market/pull/201/commits/8ce1738aae56004d9f646e692f74e1755de1815a)

ehgud0670 replied:
네임스페이스로 묶어서 사용하는 방법이 무엇인지 잘 몰라서 그러는데 혹시 예시를 보여주실 수 있을까요?!

ps. 그리고 모델이 연산 프로퍼티를 가지는건 프로퍼티처럼 소유하는건 아니고 매번 접근될때마다 계산해서 반환하므로 저는 메소드의 성격이 더 있다고 생각합니다 :)

ehgud0670 commented on OpenMarket/OpenMarket/Controller/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,134 @@
+
+final class ProductDetailViewController: UIViewController {
+    var productDetailView: ProductDetailView?
+    var productId: Int?
+    var viewControllerTitle: String?
```
이런 값들도 상위 모델을 만들어 상위 모델의 프로퍼티 값으로 위치시키는 것이 좋습니다 :)

wongbingg replied:
viewController에 전달될 정보들이 향후 많아질 수도 있다는 확장성을 고려해보니 상위모델로 묶어주는 것이 훨씬 활용성이 좋을 것 같습니다 ! ☺️☺️

ehgud0670 commented on OpenMarket/OpenMarket/Controller/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,134 @@
+                }
+                NetworkManager.shared.requestProductDeleteKey(id: productId) { key in
+                    NetworkManager.shared.requestProductDelete(id: productId, key: key) { detail in
+                        print(detail)
+                        DispatchQueue.main.async {
+                            self.showAlert(title: "삭제 성공!", message: "메인메뉴로"){
+                                self.navigationController?.popViewController(animated: true)
+                                NotificationCenter.default.post(name: .refresh, object: nil)
+                            }
+                        }
+                    }
```
저번에 말씀드련던 것처럼 뷰 업데이팅하는 메소드는 따로 분리하시면 좋을 것 같아요

wongbingg replied:
DispatchQueue.main.async 부분을 말씀하시는 걸까요 ? 이부분 마찬가지로 메서드 분리 해주었습니다 ! 🧐☺️

ehgud0670 commented on OpenMarket/OpenMarket/Controller/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,134 @@
+        let deleteCheckAlert = UIAlertController(title: "삭제", message: "비밀번호를 입력하십시오", preferredStyle: .alert)
+        deleteCheckAlert.addTextField()
+        deleteCheckAlert.addAction(UIAlertAction(title:  "입력", style: .default, handler: { UIAlertAction in
+            guard let inputKey = deleteCheckAlert.textFields?.first?.text else {
```
핸들러의 내용이 너무 긴데 메소드로 따로 빼면 좋을 것 같아요~

ehgud0670 reviewed (commented):
좋은 질문들이었어요! 마지막까지 고생하셨습니다 :)
혹시 모르거나 의문드는 부분 있으시면 말씀주세요~

ehgud0670 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -207,15 +252,45 @@ extension MainViewController {
+        self.refresher.tintColor = UIColor.red
+        self.refresher.addTarget(self, action: #selector(loadData), for: .valueChanged)
+        self.collectionView.refreshControl = refresher
+    }
+    
+    
+    
+    private func stopRefresher() {
```
```suggestion
    }
    
    private func stopRefresher() {
```
=> 여기도 메소드 간 간격은 한칸으로 해주세요~ :)

wongbingg replied:
엇. . 실수가 있었군요 ..! 꼼꼼히 봤어야했는데 ,, ㅜㅜ 수정 했습니다 !!

ehgud0670 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -207,15 +252,45 @@ extension MainViewController {
+    }
+    
+    func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
+        if collectionView.contentOffset.y > collectionView.contentSize.height - collectionView.bounds.size.height {
```
이것도 함수로 표현해주는 것이 코드파악에 좋습니다 :) 
ex) `if collectionView.isBouncingBottom`

wongbingg replied:
collectionView의 extension을 통해 긴 식을 간단하게 표현할 수 있어 좋은 것 같습니다 ! 🧐🧐

yusw10 committed:
refactor: 불필요한 print() 메서드 제거

- 개발시에 작성한 디버깅용 print 메서드 제거

yusw10 committed:
refactor: UIRefreshControl의 암시적 추출 옵셔널 해제

- MainViewController의 UIRefreshController를 옵셔널로 변경하여 암시적 추출 옵셔널을 해제함
- UIRefreshController를 사용하는 영역에서 옵셔널 체이닝으로 접근하도록 변경

yusw10 committed:
refactor: shouldHideListLayout의 기본값 설정

- 리스트와 그리드뷰 선택을 하기 위해 사용하는 shouldHideListLayout에 기본값(List 레이아웃)인 false 지정

wongbingg committed:
refactor: shouldHideListLayout 프로퍼티 내 옵저버의 메서드를 분리

- didSet 내에 메서드 한줄로 표현

wongbingg committed:
refactor: currentMaximumPage 프로퍼티 위치 변경 및 메서드네이밍 수정

- 모델에 관련된 값이므로, ProductListManager에서 관리하도록 수정
- addData() 에서 fetchMoreData() 로 네이밍을 변경

yusw10 committed:
refactor: Model 변경시점에서 collectionView가 Reload되도록 수정

- collectioniView를 직접 reload하는 방식에서 모델에 등록한 옵저버를 통해 모델값이 변경될때 리로드되도록 수정

yusw10 committed:
refactor: push 로직 분리

- DetailViewController로 넘어가는 로직을 didSelectItemAt 메서드에서 분리

yusw10 committed:
refactor: Model의 프로퍼티를 사용하는 view의 Text값을 연산프로퍼티로 정의

- ProductDetail 타입을 활용하여 정의되는 텍스트값을 모델의 연산프로퍼티로 정의하여 사용하도록 수정

wongbingg committed:
refactor: ProductDetailView 내부 update메서드 추가

- 모델을 이용하여 값들을 update해주는 로직을 DetailViewController 에서 DetailView 로 이동

wongbingg committed:
refactor: viewcontrollerTitle과 productId 를 담아줄 상위모델 생성

- 반복되서 사용되는 viewcontroller의 프로퍼티들을 상위모델로 묶어주었다

wongbingg committed:
refactor: 메서드 분리, extension으로 Bool값 처리

- ProductDetailViewController 에서, 삭제가 수행되는 메서드에 중첩되는 핸들러를 분리
- collectionView의 프로퍼티로 계산되는 Bool값을 extension을 이용해 다른곳에서 처리

ehgud0670 reviewed (approved):
피드백 반영한 것들 모두 확인했습니다. 
제가 추가로 재질문 한 것만 답변주세요!
그럼 머지할게요. 두 분 모두 고생하셨습니다~ 👏👏👏
